### PR TITLE
Resolve xcodebuild output hangs

### PR DESCRIPTION
**Pull request by Jake-Prickett**
Resolves [https://github.com/tuist/tuist/issues/2250](https://github.com/tuist/tuist/issues/2250)
Request for comments document (if applies):

### Short description 📝

Initial attempt at resolving the issue! 🚀

Currently when calling `tuist test` or `tuist build`, during compilation the output hangs indefinitely when a warning or error is encountered. The reason being, the library `xcbeautify` leveraged to format the `xcodebuild` output was designed to have output piped into it (as opposed to calling the format methods explicitly from the library). With this design, the Library calls `readLine()` when an error or warning is encountered to consume and format a bit more information about the error/warning - because Tuist does not pipe the output, `readLine()` hangs indefinitely waiting for input from `stdin`.

### Solution

Rather than format each line individually, leverage Foundation `Process` and `Pipe` to Pipe output from `xcodebuild` into `xcbeautify`. 

### Changes

- Remove `xcbeautify` as a dependency from the project
- Update the `Rakefile` to pull in the repository and build the binary version of `xcbeautify`
- Add ability to find the `xcbeautify` binary to the `BinaryLocator`
- Refactor the `Parsing` class to be a `Formatter` class that provides arguments for the `xcbeautify` command.
- `XcodeBuildController` - when running a command, we now need to know if we should pipe the output through the formatter
    - Based on `.isVerbose` from the current environment
- `System` - Created additional method to enable passing a second set of Arguments to pipe the output from the first Argument to.
    - Leverage `Foundation.Process` vs. `swift-tools-support-core` s Thread
- Adjusted the `XcodeBuildOutput` to now only have a raw string, this is due to the fact that we no longer have access to both the formatted version and the raw version at the same time.
- Updated some tests to ensure they stayed passing

### TODO:
- Finalize implementation based on feedback and make sure I followed standards :)
- Add unit test coverage for the changes added

### Checklist ✅

- [x]  The code architecture and patterns are consistent with the rest of the codebase.
- [x]  A developer other than the author has verified that the changes work as expected.
- [x]  The changes have been tested following the [documented guidelines](https://tuist.io/docs/contribution/testing-strategy/).
- [x]  The `CHANGELOG.md` has been updated to reflect the changes. In case of a breaking change, it's been flagged as such.
- [ ]  In case the PR introduces changes that affect users, the documentation has been updated.

### CODE REVIEW

**Comment by pepibumur** [comment]
Hey @Jake-Prickett, thank you very much for contributing to the project. Few suggestions:
- Rebase from `main` to fix the `Next Website`  and `Website` failing workflows.
- Run `bundle exec rake style_correct` to fix all the styling issues.
- There seems to be some legit failing unit tests.

Moreover, I'd update the CHANGELOG to mention that we are fixing a known issue running `tuist test` and `tuist test` 🚀

**Review by pepibumur — APPROVED** [review]
Hey @Jake-Prickett. Thank you very much for contributing to the project.

**Comment by pepibumur** [comment]
@all-contributors add @Jake-Prickett  for code

**Comment by Jake-Prickett** [comment]
Thanks for the review @pepibumur and the tips! Up to date with the most recent `main`, resolved formatting issues, and fixed the test failure! 👍

File: Rakefile [hunk 1]
```diff
@@ -80,6 +81,30 @@ task :swift_lint_update do
...
+      # Copy binary
+      binary_path = File.join(release_dir, "xcbeautify")
+      File.delete(dst_binary_path) if File.exist?(dst_binary_path)
+      FileUtils.cp(binary_path, dst_binary_path)
+    end
+  end
+  # Write version
+  File.write(File.join(root_dir, "vendor/.xcbeautify.version"), XCBEAUTIFY_VERSION)
+end
+
```
Comment (RomainBoulay) [inline Rakefile:107]: Nice 👍

File: CHANGELOG.md [hunk 2]
```diff
@@ -12,6 +12,7 @@ Please, check out guidelines: https://keepachangelog.com/en/1.0.0/
 
 - Update failing trying to create the `swift-project` symbolic link [#2244](https://github.com/tuist/tuist/pull/2244)
 - Tuist now correctly parses arm64e architectures in xcframeworks [#2247](https://github.com/tuist/tuist/pull/2247) by [@thedavidharris](https://github.com/thedavidharris).
+- Fixed known issue when running `tuist test` and `tuist build`. [#2297](https://github.com/tuist/tuist/pull/2297) by [@Jake-Prickett](https://github.com/Jake-Prickett).
```
Comment (RomainBoulay) [inline CHANGELOG.md:15]:
```suggestion
- Fixed known issue that makes the process hang when running `tuist test` and `tuist build`. [#2297](https://github.com/tuist/tuist/pull/2297) by [@Jake-Prickett](https://github.com/Jake-Prickett).
```

It's just a minor suggestion! Also English isn't my first language so feel free to ignore or rephrase!

**Review by RomainBoulay — APPROVED** [review]
Very solid work @Jake-Prickett 👏 Thanks a lot for jumping on that 🙏

File: CHANGELOG.md (hunk 2 above)
Comment (Jake-Prickett) [reply CHANGELOG.md]: I like the suggestion! Wasn't sure how in depth to get. 👍🏼

File: Sources/TuistAutomation/Utilities/Formatter.swift [hunk 3]
```diff
@@ -0,0 +1,20 @@
+import Foundation
+import TuistCore
+import TuistSupport
+
+public protocol Formatting {
```
Comment (fortmarek) [inline Sources/TuistAutomation/Utilities/Formatter.swift:5]: I don't think this has to be public ... or am I mistaken?

**Review by fortmarek — APPROVED** [review]
Good catch, found one small thing (maybe), thanks 👏

File: Sources/TuistAutomation/Utilities/Formatter.swift (hunk 3 above)
Comment (Jake-Prickett) [reply Sources/TuistAutomation/Utilities/Formatter.swift]: You are correct. Great catch, thank you! 👍

**Comment by fortmarek** [comment]
The unit tests seem to have failed, otherwise I think we are good to merge this

**Comment by Jake-Prickett** [comment]
To be honest, I'm not entirely sure what the problem is. 🤔 

It seems as though the 11.5 one is the issue if I'm not mistaken? Which led me to thinking it could be a difference with how `xcbeautify` may parse the output from `xcodebuild` for Xcode 11.5 (maybe a regex doesn't match or something?) 

But when running locally (using both Xcode and command line) the tests that seem to fail/hang on CI pass without any issues on both Xcode 11.5 & Xcode 12.3. 

**Running the test that hangs:**
```shell
➜  tuist git:(resolve-xcodebuild-output-hangs) swift package clean
➜  tuist git:(resolve-xcodebuild-output-hangs) swift test --filter TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests
# ... compilation warnings, etc.
[1822/1822] Linking tuistPackageTests
Test Suite 'Selected tests' started at 2021-01-16 09:49:14.007
Test Suite 'tuistPackageTests.xctest' started at 2021-01-16 09:49:14.007
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' started at 2021-01-16 09:49:14.007
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_iOS_framework]' started.
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_iOS_framework]' passed (7.622 seconds).
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' passed at 2021-01-16 09:49:21.630.
	 Executed 1 test, with 0 failures (0 unexpected) in 7.622 (7.623) seconds
Test Suite 'tuistPackageTests.xctest' passed at 2021-01-16 09:49:21.630.
	 Executed 1 test, with 0 failures (0 unexpected) in 7.622 (7.623) seconds
Test Suite 'Selected tests' passed at 2021-01-16 09:49:21.630.
	 Executed 1 test, with 0 failures (0 unexpected) in 7.622 (7.624) seconds
Test Suite 'Selected tests' started at 2021-01-16 09:49:21.707
Test Suite 'tuistPackageTests.xctest' started at 2021-01-16 09:49:21.708
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' started at 2021-01-16 09:49:21.708
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_macOS_framework]' started.
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_macOS_framework]' passed (3.211 seconds).
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' passed at 2021-01-16 09:49:24.919.
	 Executed 1 test, with 0 failures (0 unexpected) in 3.211 (3.212) seconds
Test Suite 'tuistPackageTests.xctest' passed at 2021-01-16 09:49:24.919.
	 Executed 1 test, with 0 failures (0 unexpected) in 3.211 (3.212) seconds
Test Suite 'Selected tests' passed at 2021-01-16 09:49:24.919.
	 Executed 1 test, with 0 failures (0 unexpected) in 3.211 (3.212) seconds
Test Suite 'Selected tests' started at 2021-01-16 09:49:24.970
Test Suite 'tuistPackageTests.xctest' started at 2021-01-16 09:49:24.970
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' started at 2021-01-16 09:49:24.970
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_tvOS_framework]' started.
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_tvOS_framework]' passed (5.788 seconds).
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' passed at 2021-01-16 09:49:30.759.
	 Executed 1 test, with 0 failures (0 unexpected) in 5.788 (5.789) seconds
Test Suite 'tuistPackageTests.xctest' passed at 2021-01-16 09:49:30.759.
	 Executed 1 test, with 0 failures (0 unexpected) in 5.788 (5.789) seconds
Test Suite 'Selected tests' passed at 2021-01-16 09:49:30.759.
	 Executed 1 test, with 0 failures (0 unexpected) in 5.788 (5.790) seconds
Test Suite 'Selected tests' started at 2021-01-16 09:49:30.830
Test Suite 'tuistPackageTests.xctest' started at 2021-01-16 09:49:30.831
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' started at 2021-01-16 09:49:30.831
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_watchOS_framework]' started.
Test Case '-[TuistKitIntegrationTests.CacheXCFrameworkBuilderIntegrationTests test_build_when_watchOS_framework]' passed (6.366 seconds).
Test Suite 'CacheXCFrameworkBuilderIntegrationTests' passed at 2021-01-16 09:49:37.198.
	 Executed 1 test, with 0 failures (0 unexpected) in 6.366 (6.367) seconds
Test Suite 'tuistPackageTests.xctest' passed at 2021-01-16 09:49:37.198.
	 Executed 1 test, with 0 failures (0 unexpected) in 6.366 (6.367) seconds
Test Suite 'Selected tests' passed at 2021-01-16 09:49:37.198.
	 Executed 1 test, with 0 failures (0 unexpected) in 6.366 (6.368) seconds
```

Any suggestions on how to approach this one? It looks like the run hung for ~3 hours, I don't want to eat up CI resources for that long 🤯

**Comment by Jake-Prickett** [comment]
### Update:

I'm able to run the Unit Tests locally on both Xcode 11.5 and 12.3 without having any issues. Additionally, when running the Acceptance Tests (leveraging `./bin fourier test tuist acceptance ...`) they pass as well. 🤔 

It doesn't look to be related to `xcbeautify` from my most recent investigation. I feel like I'm missing something small, but hit a wall with debugging this one. I'm trying to avoid debugging on CI by pushing additional commits since it blocks the CI resources for quite some time. 🤯 Any suggestions or tips @pepibumur @RomainBoulay @fortmarek?

**Comment by pepibumur** [comment]
I'm looking into this one

**Comment by pepibumur** [comment]
@Jake-Prickett I moved your work to this PR so that I can debug and push a fix to the branch. I also invited you to the organization so that you can push to remote without having to fork the repo.

**Comment by Jake-Prickett** [comment]
Thank you @pepibumur! 🙏 I'll continue debugging on my end as well!

**Pull request closed without merging**